feat(event_source): Event source dataclasses for IoT Core Registry Events

basvandriel opened the pull request:
<!-- markdownlint-disable MD041 MD043 -->
**Issue number:** #6110 

## Summary

### Changes

On top of the pydantic classes for the IoT registry events, this pull request also implements the event source dataclasses.

### User experience

```python
from aws_lambda_powertools.utilities.data_classes import event_source
from aws_lambda_powertools.utilities.data_classes.iot_registry_event import IoTCoreThingEvent


@event_source(data_class=IoTCoreThingEvent)
def lambda_handler(event: IoTCoreThingEvent, context):
    print(f"Received IoT Core event type {event.event_type}")
```

## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [x] [Meet tenets criteria](https://docs.powertools.aws.dev/lambda/python/#tenets)
* [x] I have performed a self-review of this change
* [x] Changes have been tested
* [x] Changes are documented
* [x] PR title follows [conventional commit semantics](https://github.com/aws-powertools/powertools-lambda-python/blob/develop/.github/semantic.yml)

<details>
<summary>Is this a breaking change?</summary>

**RFC issue number**:

Checklist:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

</details>

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

basvandriel committed:
Comment out fields to work with

basvandriel committed:
work on tests

basvandriel committed:
Add docs

basvandriel commented:
Closes https://github.com/aws-powertools/powertools-lambda-python/issues/6110

basvandriel committed:
Add initial doc links

basvandriel committed:
work

basvandriel committed:
fix links

basvandriel committed:
Add github links

basvandriel committed:
work

basvandriel committed:
Merge branch 'develop' into feature/event-source-iot-registry-events

basvandriel commented:
Hi @leandrodamascena, would you like me to add examples for every class? I already added the documentation headers, but was not sure to already implement them all since their quite identical.

leandrodamascena commented on aws_lambda_powertools/utilities/data_classes/iot_registry_event.py:
```diff
@@ -0,0 +1,418 @@
+        return self["versionNumber"]
+
+    @property
+    def thing_type_name(self) -> Optional[str]:
```
```suggestion
    def thing_type_name(self) -> str | None:
```

leandrodamascena commented on aws_lambda_powertools/utilities/data_classes/iot_registry_event.py:
```diff
@@ -0,0 +1,418 @@
+        return self.get("thingTypeName")
+
+    @property
+    def attributes(self) -> Dict[str, Any]:
```
```suggestion
    def attributes(self) -> dict[str, Any]:
```

leandrodamascena commented on aws_lambda_powertools/utilities/data_classes/iot_registry_event.py:
```diff
@@ -0,0 +1,418 @@
+        return self["isDeprecated"]
+
+    @property
+    def deprecation_date(self) -> Optional[datetime]:
```
```suggestion
    def deprecation_date(self) -> datetime | None:
```

leandrodamascena commented on aws_lambda_powertools/utilities/data_classes/iot_registry_event.py:
```diff
@@ -0,0 +1,418 @@
+        return datetime.fromisoformat(self["deprecationDate"]) if self.get("deprecationDate") else None
+
+    @property
+    def searchable_attributes(self) -> List[str]:
```
```suggestion
    def searchable_attributes(self) -> list[str]:
```

leandrodamascena commented on aws_lambda_powertools/utilities/data_classes/iot_registry_event.py:
```diff
@@ -0,0 +1,418 @@
+        return self["searchableAttributes"]
+
+    @property
+    def propagating_attributes(self) -> List[Dict[str, str]]:
```
```suggestion
    def propagating_attributes(self) -> list[dict[str, str]]:
```

leandrodamascena commented on aws_lambda_powertools/utilities/data_classes/iot_registry_event.py:
```diff
@@ -0,0 +1,418 @@
+        return self["versionNumber"]
+
+    @property
+    def parent_group_name(self) -> Optional[str]:
```
```suggestion
    def parent_group_name(self) -> str | None:
```

leandrodamascena commented on aws_lambda_powertools/utilities/data_classes/iot_registry_event.py:
```diff
@@ -0,0 +1,418 @@
+        return self.get("parentGroupName")
+
+    @property
+    def parent_group_id(self) -> Optional[str]:
```
```suggestion
    def parent_group_id(self) -> str | None:
```

leandrodamascena commented on aws_lambda_powertools/utilities/data_classes/iot_registry_event.py:
```diff
@@ -0,0 +1,418 @@
+        return self["description"]
+
+    @property
+    def root_to_parent_thing_groups(self) -> List[Dict[str, str]]:
```
```suggestion
    def root_to_parent_thing_groups(self) -> list[dict[str, str]]:
```

leandrodamascena commented on aws_lambda_powertools/utilities/data_classes/iot_registry_event.py:
```diff
@@ -0,0 +1,418 @@
+        return self["attributes"]
+
+    @property
+    def dynamic_group_mapping_id(self) -> Optional[str]:
```
```suggestion
    def dynamic_group_mapping_id(self) -> str | None:
```

leandrodamascena commented on aws_lambda_powertools/utilities/data_classes/iot_registry_event.py:
```diff
@@ -0,0 +1,418 @@
+        return self["rootToParentThingGroups"]
+
+    @property
+    def attributes(self) -> Dict[str, Any]:
```
```suggestion
    def attributes(self) -> dict[str, Any]:
```

leandrodamascena commented on aws_lambda_powertools/utilities/data_classes/iot_registry_event.py:
```diff
@@ -0,0 +1,418 @@
+from __future__ import annotations
+
+from datetime import datetime
+from typing import Any, Dict, List, Literal, Optional
```
```suggestion
from typing import Any, Literal
```

leandrodamascena reviewed (changes requested):
Hi @basvandriel! SUPER THANKS for another super PR! I just left some comments to adhere to this [PEP](https://peps.python.org/pep-0563/).

If possible, please add a small example for each class. If you don't have time, please let me know that I can do this.

basvandriel commented:
> Hi @basvandriel! SUPER THANKS for another super PR! I just left some comments to adhere to this [PEP](https://peps.python.org/pep-0563/).
> 
> If possible, please add a small example for each class. If you don't have time, please let me know that I can do this.

Hi @leandrodamascena, will do! Thanks for the review. I'll try to find time to implement the examples this week.

basvandriel committed:
add files

basvandriel committed:
fill classes

basvandriel committed:
add log statement

leandrodamascena committed:
Merge branch 'develop' into feature/event-source-iot-registry-events

basvandriel committed:
Merge branch 'develop' into feature/event-source-iot-registry-events

basvandriel committed:
work

basvandriel committed:
apply suggestions

basvandriel committed:
Add examples for first items

basvandriel committed:
Fix documentation

basvandriel committed:
apply suggestion

basvandriel committed:
Add doc example

basvandriel committed:
Work

basvandriel committed:
work

leandrodamascena commented:
Hey @basvandriel, thanks for accepting the suggestions and working to address them! I'm currently away from my computer, but I'll review everything tomorrow morning. 

Everything looks good, but If any small changes are needed, I'll send a small commit and merge it. I'd like to include it in the release we're planning for tomorrow.

Thanks a lot for this another super PR! 🎉🎉👏👏

basvandriel commented:
Hi @leandrodamascena, again, sorry for the late reply. Feel free to review my latest changes, I've updated documentation and added examples as discussed. I had to commit your suggested changes manually, as GitHub wouldn't let me do it in the browser. Anyway - looking forward to your response!

leandrodamascena committed:
Merge branch 'develop' into feature/event-source-iot-registry-events

leandrodamascena committed:
Adding more coverage + documentation

leandrodamascena committed:
Adding more coverage + documentation

leandrodamascena committed:
Adding more coverage + documentation

leandrodamascena committed:
Adding more coverage + documentation

leandrodamascena committed:
Adding more coverage + documentation

leandrodamascena reviewed (approved):
Hi @basvandriel! Thanks a lot for working on this PR! I just made some minor changes to the documentation and added all the fields to the test file to increase coverage.

APPROVED! This will be included in the release in the next few hours!
